sheep/src/pack/maxrects.rs

```diff
@@ -0,0 +1,369 @@
+                break;
+            }
+
+            placeable.sort_by_key(|(score, _)| score.primary);
+            let (score, sprite) = {
+                // Find out if there's multiple with the best score
+                let best_scored = placeable
+                    .iter()
+                    .filter(|(score, _)| score.primary == placeable[0].0.primary)
+                    .collect::<Vec<&(RectScore, SpriteData)>>();
+
+                // If not, we have found the next best fit! Othweise, take the
+                // next best by the secondary score
+                if best_scored.len() == 1 {
+                    best_scored[0]
+                } else {
+                    best_scored
+                        .iter()
+                        .min_by_key(|(score, _)| score.secondary)
+                        .expect("Unreachable")
+                }
+            };
+
+            self.place_rect(score.placement, sprite.id);
+            sprites.retain(|s| s.id != sprite.id);
+            placed.push(sprite.id);
+        }
+
+        sprites
+    }
+
+    pub fn score_rect(&self, width: u32, height: u32) -> ScoreResult {
+        use std::cmp::{max, min};
+
+        // We score by best short side fit, since it's the best performing
+        // strategy according to the reference implementation
+        let mut best_short = std::u32::MAX;
+        let mut best_long = std::u32::MAX;
+        let mut placement = Rect::new(0, 0, 0, 0);
+
+        self.free.iter().for_each(|it| {
+            let leftover_horiz = (it.width as i32 - width as i32).abs() as u32;
+            let leftover_vert = (it.height as i32 - height as i32).abs() as u32;
+
+            let short_side_fit = min(leftover_horiz, leftover_vert);
+            let long_side_fit = max(leftover_horiz, leftover_vert);
+
+            if short_side_fit < best_short
+                || (short_side_fit == best_short && long_side_fit < best_long)
+            {
+                best_short = short_side_fit;
+                best_long = long_side_fit;
+                placement = Rect::new(it.x, it.y, width, height);
+            }
+        });
+
+        // TODO(happenslol): This is kind of a primitive way to check,
+        // since it's directly translated from the reference implementation.
+        // This function can probably be improved a lot, style-wise
+        if placement.height == 0 {
+            ScoreResult::NoFit
+        } else {
+            ScoreResult::FitFound(RectScore {
+                placement,
+                primary: best_short as i32,
+                secondary: best_long as i32,
+            })
+        }
+    }
+
+    fn place_rect(&mut self, rect: Rect, sprite_id: usize) {
+        let mut to_process = self.free.len();
+        let mut i = 0;
+
+        while i < to_process {
+            if self.free[i].no_intersection(&rect) {
+                i += 1;
+                continue;
+            }
+
+            let to_split = self.free.remove(i);
```

Comment:
this can probably be `swap_remove`

Reply:
Since i'm looping, this would complicate things imo

Reply:
i don't think so. You remove and revisit the same index later. If order doesn't matter, it should all stay the same.